src/basic/cgroup-util.c

```diff
@@ -1198,6 +1198,34 @@ int cg_path_get_unit(const char *path, char **ret) {
         return 0;
 }
 
+int cg_path_get_unit_path(const char *path, char **ret) {
+        char *path_copy, *unit_name, *unit_path;
+        size_t unit_name_len;
+
+        assert(path);
+        assert(ret);
+
+        path_copy = strdupa_safe(path);
+
+        unit_name = (char *)skip_slices(path_copy);
+        unit_name_len = strcspn(unit_name, "/");
+        if (unit_name_len < 3)
+                return -ENXIO;
+
+        unit_name[unit_name_len] = '\0';
+
+        if (!unit_name_is_valid(unit_name, UNIT_NAME_PLAIN|UNIT_NAME_INSTANCE))
```

Comment:
now you dropped the unescaping entirely. 

you should do the unescaping here, but only for the validation step, and not modify the path string.

i.e. it's sufficient to change this line here:

```c
if (!unit_name_is_valid(cg_unescape(unit_name), UNIT_NAME_PLAIN|UNIT_NAME_INSTANCE))
```